presto-spi/src/main/java/io/prestosql/spi/block/DictionaryBlock.java

```diff
@@ -497,20 +497,15 @@ private DictionaryBlock unnest()
             return this;
         }
 
+        DictionaryBlock nestedDictionary = (DictionaryBlock) dictionary;
+        nestedDictionary = nestedDictionary.unnest();
+        dictionary = nestedDictionary;
```

Comment:
Previously, `unnest()` (and methods calling it like `compact()`, `calculateCompactSize()`) were read-only (except of `sizeInBytes` and `uniqueIds` which is safe to write without any synchronization).

After the change, `compact()` both: returns a compact block _and_ changes internal state.

Reply:
dictionaries are unnested anyway. We just make sure they get memoized, so they won't get unnested again. This operation is idempotent

Reply:
The class intends to be thread-safe (otherwise there would not be `volatile`) and I think it should be.
It's not safe to change a (generally) read-only method into a read-write one. If the block is read by a concurrent thread, replacing `dictionary` will break that.